pam/internal/adapter/nativemodel.go

```diff
@@ -584,13 +584,13 @@ func (m nativeModel) handleFormChallenge(hasWait bool) tea.Cmd {
 		}
 	}
 
-	instructions := "Insert '%[1]s' to cancel the request and go back"
+	instructions := "Type '%[1]s' to cancel the request and go back"
 	if hasWait {
 		// Duplicating some contents here, as it will be better for translators once we've them
-		instructions = "Leave the input field empty to wait for other authentication method " +
+		instructions = "Leave the input field empty to wait for the alternative authentication method " +
```

Comment:
What is the "input field" here btw? Isn't the native model for usage on the command-line?

Reply:
And how does the UI differ when `m.uiLayout.GetEntry() == ""`?

Reply:
Well, in case we've a `wait` form, we have two options:
 - A case where we only ask for the user to start waiting for another device to authenticate
 - A case where the user may both insert some data in the input field (like a pin-code) or to wait for another authentication method (like clicking on a link...)

Reply:
You can see examples in the MFA golden files, where one may just wait for something or give the user the option to insert some challenge.

Reply:
I'm looking at https://github.com/3v1n0/authd/blob/5de61a1fc578cbc0d946e5e8770cb290c581b404/pam/integration-tests/testdata/TestNativeAuthenticate/golden/authenticate_user_switching_auth_mode#L81-L84 but I still don't understand which "alternative authentication method" this string is referring to.

I think just appending ` (or enter 'r' to choose another authentication method)` to the prompt would work here as well (same as in https://github.com/ubuntu/authd/pull/599/files#r1810665234):

```
Click on the link received at user-integration-switch-mode@gmail.com or enter the code (or enter 'r' to choose another authentication method):
```

WDYT?

Reply:
Well, I didn't like the idea of sending a very long string as the input field text as otherwise we'll get the line you posted and then the user has to type inline.

So was thinking to just keep the broker-provided field as the input field, while the rest is just some introduction to that, not to make the actual broker-provided string too hard to read in long cases, but I've not very strong opinions here.

Maybe @didrocks has ([this is a golden](https://github.com/ubuntu/authd/blob/main/pam/integration-tests/testdata/TestNativeAuthenticate/golden/authenticate_user_switching_auth_mode) we're trying to change)?

Reply:
This is what it looks like when logging in with SSH on this branch and going back to the provider menu:

```
ssh foo@bar.onmicrosoft.com@localhost
Enter 'r' to cancel the request and go back
(foo@bar.onmicrosoft.com@localhost) Enter your local password: 
(foo@bar.onmicrosoft.com@localhost) == Authentication method selection (enter 'r' to go back) ==
  1. Local Password Authentication
  2. Device Authentication
Choose your authentication method: r
(foo@bar.onmicrosoft.com@localhost) == Provider selection ==
  1. local
  2. Microsoft Entra ID
Choose your provider: 
```

This is a proposal with some changes to the prompts which IMO provides better UX:

```
ssh foo@bar.onmicrosoft.com@localhost
(foo@bar.onmicrosoft.com@localhost) == Local Password Authentication ==
Enter your local password (or enter 'r' to choose another authentication method):
> 
(foo@bar.onmicrosoft.com@localhost) == Authentication method selection ==
  1. Local Password Authentication
  2. Device Authentication (Microsoft Entra ID)
Choose your authentication method (or enter 'r' to choose another provider):
> r
(foo@bar.onmicrosoft.com@localhost) == Provider selection ==
  1. local
  2. Microsoft Entra ID
Choose your provider:
> 
```

The changes are:
* The alternative entry (`'r'`) is placed in parentheses directly after the actual prompt. This makes it easier to understand that the alternative entry should be made in the same prompt.
* The entry is made in a separate line, prefixed with `> `, to avoid inline entry in an overly long line
  * We already do that when prompting for the local password in the `login` command and IMO it looks good (even though IMO it still hurts UX that echo is off, but that's how password input usually works on the command-line).
* We tell the user what they can do when they type 'r', instead of the vague "to go back" (back where? when logging in with SSH, they didn't go through that menu in the first place)
* We show a header `== Local Password Authentication ==` when asking for the local password, which is consistent with the other prompts which allow the user to go back, and conveys the impression of being in a multi-step dialog instead of a simple prompt.

Reply:
Nice ideas, I'll get through them!

Reply:
So... This works fine when used in terminal interfaces, but I'm not sure about other UIs (like polkit):

![image](https://github.com/user-attachments/assets/adba94d2-0f43-4b5a-b517-302404a2b745)


![image](https://github.com/user-attachments/assets/7f1a82c7-1559-419f-9b3c-91e0295d2ed1)

Compared to previous state:

![image](https://github.com/user-attachments/assets/38c71803-0c94-44b5-a72f-3eee3390566b).

So probably in the polkit case it's better to split the info messages with the prompt one so that the UI keeps working as before, or let me think better...

Reply:
And the case of qrcode is even more complex because we need to notify infos, that otherwise we'd loose (this is with the old code, moving things inline may break this):

![image](https://github.com/user-attachments/assets/103f9d3f-1386-4e55-a515-b8f4fa7c01b0)

Reply:
Oh boy, the UX with polkit is so weird. If we can't at least change the button of the "Authenticate" label, I think we should reconsider if it's even worth to support going "back" in polkit.

Reply:
Well, polkit UI indeed needs love, but to get there there are too many layers to cover for now, so I wouldn't mind on that, but at least to show the proper values.

So maybe we should just rely on having the info messages as something separated from the actual prompt. So that they get properly handled by all the potential PAM UIs

Reply:
So, I've changed a bit of things so that everything should be readable and working both for polkit and SSH:

![image](https://github.com/user-attachments/assets/a01a50ad-46f1-4934-94de-221b4f16ff95)
![image](https://github.com/user-attachments/assets/4e2d0498-e3ce-46e9-84f7-3d61de626232)
![image](https://github.com/user-attachments/assets/8c7dc919-0d20-4421-ab3b-fde786462724)

While the the ssh is:

```
$ ssh $USER@localhost
== Qr Code authentication ==
Scan the qrcode or enter the code in the login page
https://ubuntu.com
       1337       
 
  1. Wait for the QR code scan result
  2. Regenerate code
Or enter 'r' to go back to select the authentication method
(user-sudo@localhost) Choose action:
> 
```

or

```
== Authentication method selection ==
  1. Use a QR code
  2. Password authentication
  3. Send URL to user-sudo@gmail.com
  4. Use your fido device foo
  5. Use your phone +33\u2026
  6. Use your phone +1\u2026
  7. Pin code
  8. Authentication code
Or enter 'r' to go back to choose the provider
(user-sudo@localhost) Choose your authentication method:
> 
```

```
== Password authentication ==
Enter 'r' to cancel the request and go back to select the authentication method
(user-sudo@localhost) Gimme your password:
> 
```

Reply:
It's possible (in another MR and after discussing what are the potential drawbacks for the upcoming brokers), but in any case we still need to use the very same logic to ensure that both the prompt and the info messages are shown.

So I'd focus on the look on the terminal now, but keeping in mind that for now we've to support both cases.

Reply:
> It's possible (in another MR and after discussing what are the potential drawbacks for the upcoming brokers),

Ok, I'll file a ticket for that

> any case we still need to use the very same logic to ensure that both the prompt and the info messages are shown

but if we don't have to consider the pkexec use case, we could use the inline style (`Enter your local password (or enter 'r' to choose another authentication method):`) again, right? 

We could facilitate translations by letting the translators decide where the prompt placeholder should be:

```go
// Translators: %s will be replaced with the prompt, for example "Enter your local password"
prompt := fmt.Sprintf("%s (or enter 'r' to choose another authentication method)", prompt)
```

Reply:
> No if we don't want to break it (since the message won't be visible)

I understand that the message is not visible in pkexec, that's why I assume that if we don't have to support the pkexec use case, we could use the inline style again. Or are there other UIs in which a longer prompt message won't be visible?

> But in any case, I think it's not correct to use very long prompt messages, neither if you want to use this with lightdm for example (that used to work, more or less). Or a CLI UI in general, since that's what the prompt should ask for, while the rest of information is likely to be going into info messages.

The prompt message should describe what we expect the user to enter in the prompt. Since we support both entering a password (in the password case) and 'r', we should mention both in the prompt message. That's why I think that

```
(foo@bar.onmicrosoft.com@localhost) == Local Password Authentication ==
Enter your local password (or enter 'r' to choose another authentication method):
> 
```

still provides better UX than

```
== Local Password Authentication ==
Enter 'r' to cancel the request and go back to select the authentication method
(foo@bar.onmicrosoft.com@localhost) Enter your local password:
```

> Yeah, but you end up splitting sentences and this doesn't really work well with many languages and grammars, see for example this guideline: https://wiki.gnome.org/TranslationProject(2f)DevGuidelines(2f)Never(20)split(20)sentences.html

Translators could also make a second sentence out of the part that's in parenthesis. I see that this is not ideal and makes the work of the translators harder though. As I said elsewhere, I think that ideally all user-visible strings should be provided by the model (and not the broker), and the broker should only tell the model which kind of information it needs / can provide (password, QR code). That would allow us to put the complete sentence in one string which translators can then translate:
```go
prompt := "Enter your local password (or enter 'r' to choose another authentication method)"
```

Reply:
> But in any case, I think it's not correct to use very long prompt messages, neither if you want to use this with lightdm for example (that used to work, more or less)

Can we list somewhere which PAM UIs we want to support, and how to test them? That would make it easier to manually test changes like these.

Reply:
Well, so far the only ones we care about are `ssh` and polkit, I don't think we claim as supported anything else other than those.

But technically if an user changes the desktop environment, then we still try to work even if the experience may not be 1st class citizen.

Reply:
Then I think it would make sense to have a list of the PAM apps which we want to support + a list of other PAM apps which are nice to have but not our focus (so we can keep an eye on how things look in those but don't put much effort into improving things there).

Reply:
> Then I think it would make sense to have a list of the PAM apps which we want to support + a list of other PAM apps which are nice to have but not our focus (so we can keep an eye on how things look in those but don't put much effort into improving things there).

I filed https://github.com/ubuntu/authd/issues/618 to track that.

Reply:
Feel free to resolve this thread.